primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,494 @@
+					// write was good, but read failed
+					Some(None) => Ok(new_deadline),
+					// lock is set, but it is expired. We can re-acquire it.
+					Some(Some(deadline)) if <L as Lockable>::has_expired(&deadline) =>
+						Ok(new_deadline),
+					// lock is present and is still active
+					Some(Some(deadline)) => Err(deadline),
+				}
+			},
+		);
+		match res {
+			Ok(Ok(_)) => Ok(()),
+			Ok(Err(deadline)) => Err(deadline),
+			Err(e) => Err(e),
+		}
+	}
+
+	/// A single attempt to lock using the storage entry.
+	///
+	/// Returns a lock guard on success, otherwise an error containing the
+	/// `<Self::Lockable>::Deadline` in for the currently active lock
+	/// by another task `Err(<L as Lockable>::Deadline)`.
+	pub fn try_lock(&mut self) -> Result<StorageLockGuard<'a, '_, L>, <L as Lockable>::Deadline> {
+		self.try_lock_inner(self.lockable.deadline())?;
+		Ok(StorageLockGuard::<'a, '_> { lock: Some(self) })
+	}
+
+	/// Repeated lock attempts until the lock is successfully acquired.
+	///
+	/// If one uses `fn forget(..)`, it is highly likely `fn try_lock(..)`
+	/// is the correct API to use instead of `fn lock(..)`, since that might
+	/// never unlock in the anticipated span i.e. when used with `BlockAndTime`
+	/// during a certain block number span.
+	pub fn lock(&mut self) -> StorageLockGuard<'a, '_, L> {
+		while let Err(deadline) = self.try_lock_inner(self.lockable.deadline()) {
+			L::snooze(&deadline);
+		}
+		StorageLockGuard::<'a, '_, L> { lock: Some(self) }
+	}
+
+	/// Explicitly unlock the lock.
+	fn unlock(&mut self) {
+		self.value_ref.clear();
+	}
+}
+
+/// RAII style guard for a lock.
+pub struct StorageLockGuard<'a, 'b, L: Lockable> {
+	lock: Option<&'b mut StorageLock<'a, L>>,
+}
+
+impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
+	/// Consume the guard but **do not** unlock the underlying lock.
+	///
+	/// Can be used to implement a grace period after doing some
+	/// heavy computations and sending a transaction to be included
+	/// on-chain. By forgetting the lock, it will stay locked until
+	/// its expiration deadline is reached while the off-chain worker
+	/// can already exit.
+	pub fn forget(mut self) {
+		let _ = self.lock.take();
+	}
+}
+
+impl<'a, 'b, L: Lockable> Drop for StorageLockGuard<'a, 'b, L> {
+	fn drop(&mut self) {
+		if let Some(lock) = self.lock.take() {
+			lock.unlock();
+		}
+	}
+}
+
+impl<'a> StorageLock<'a, Time> {
+	/// Explicitly create a time based storage lock with a non-default
+	/// expiration timeout.
+	pub fn with_deadline(key: &'a [u8], expiration_duration: Duration) -> Self {
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			lockable: Time {
+				expiration_duration: expiration_duration,
+			},
+		}
+	}
+}
+
+impl<'a, B> StorageLock<'a, BlockAndTime<B>>
+where
+	B: BlockNumberProvider,
+{
+	/// Explicitly create a time and block number based storage lock with
+	/// a non-default expiration duration and block number offset.
+	pub fn with_block_and_time_deadline(
```

Comment:
Since there is a default value for time-based lock anyway, I'd suggest to add `with_block_deadline` method which uses that and only requires the block timeout to be specified.